File: linda-logic/src/main/java/it/unibo/coordination/linda/logic/LogicSpace.java
package it.unibo.coordination.linda.logic;

import alice.tuprolog.Term;
import it.unibo.coordination.Engines;
import it.unibo.coordination.linda.core.Match;
import it.unibo.coordination.linda.core.TupleSpace;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;

public interface LogicSpace extends TupleSpace<LogicTuple, LogicTemplate, String, Term> {

    static LogicSpace deterministic(String name, ExecutorService executorService) {
        return new DeterministicLogicSpaceImpl(name, executorService);
    }

    static LogicSpace deterministic(String name) {
        return deterministic(name, Engines.getDefaultEngine());
    }

    static LogicSpace deterministic(ExecutorService executorService) {
        return deterministic(null, Engines.getDefaultEngine());
    }

    static LogicSpace nonDeterministic(String name, ExecutorService executorService) {
        throw new UnsupportedOperationException("not implemented");
    }

    static LogicSpace nonDeterministic(String name) {
        return nonDeterministic(name, Engines.getDefaultEngine());
    }

    static LogicSpace nonDeterministic(ExecutorService executorService) {
        return nonDeterministic(null, Engines.getDefaultEngine());
    }


    default CompletableFuture<LogicTuple> write(String tuple) {
        return write(LogicTuple.of(tuple));
    }

    default CompletableFuture<LogicTuple> write(Term tuple) {
        return write(LogicTuple.of(tuple));
    }



    default CompletableFuture<LogicTuple> readTuple(String template) {
        return readTuple(LogicTemplate.of(template));
    }

    default CompletableFuture<LogicTuple> readTuple(Term template) {
        return readTuple(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> read(String template) {
        return read(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> tryRead(String template) {
        return tryRead(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> tryRead(Term template) {
        return tryRead(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> tryAbsent(String template) {
        return tryAbsent(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> tryAbsent(Term template) {
        return tryAbsent(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> tryTake(String template) {
        return tryTake(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> tryTake(Term template) {
        return tryTake(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> read(Term template) {
        return read(LogicTemplate.of(template));
    }

    default CompletableFuture<Optional<LogicTuple>> tryReadTuple(String template) {
        return tryReadTuple(LogicTemplate.of(template));
    }

    default CompletableFuture<Optional<LogicTuple>> tryReadTuple(Term template) {
        return tryReadTuple(LogicTemplate.of(template));
    }

    default CompletableFuture<LogicTuple> takeTuple(String template) {
        return takeTuple(LogicTemplate.of(template));
    }

    default CompletableFuture<LogicTuple> takeTuple(Term template) {
        return takeTuple(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> take(String template) {
        return take(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> take(Term template) {
        return take(LogicTemplate.of(template));
    }

    default CompletableFuture<Optional<LogicTuple>> tryTakeTuple(String template) {
        return tryTakeTuple(LogicTemplate.of(template));
    }

    default CompletableFuture<Optional<LogicTuple>> tryTakeTuple(Term template) {
        return tryTakeTuple(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> absent(final String template) {
        return absent(LogicTemplate.of(template));
    }

    default CompletableFuture<Match<LogicTuple, LogicTemplate, String, Term>> absent(final Term template) {
        return absent(LogicTemplate.of(template));
    }

    default CompletableFuture<Optional<LogicTuple>> tryAbsentTuple(String template) {
        return tryAbsentTuple(LogicTemplate.of(template));
    }

    default CompletableFuture<Optional<LogicTuple>> tryAbsentTuple(Term template) {
        return tryAbsentTuple(LogicTemplate.of(template));
    }

    @Override
    default LogicTuple toTuple(String $this$toTuple) {
        return LogicTuple.of($this$toTuple);
    }

    @Override
    default LogicTemplate toTemplate(String $this$toTemplate) {
        return LogicTemplate.of($this$toTemplate);
    }
}

File: linda-logic/src/main/kotlin/it/unibo/coordination/linda/logic/LogicSpace.kt
package it.unibo.coordination.linda.logic

import it.unibo.coordination.Engines
import it.unibo.coordination.linda.core.TupleSpace
import it.unibo.tuprolog.core.Term
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService

interface LogicSpace : TupleSpace<LogicTuple, LogicTemplate, String, Term, LogicMatch> {

    @JvmDefault
    fun write(tuple: Term): CompletableFuture<LogicTuple> = write(LogicTuple.of(tuple))

    @JvmDefault
    fun readTuple(template: Term): CompletableFuture<LogicTuple> = readTuple(LogicTemplate.of(template))

    @JvmDefault
    fun tryRead(template: Term): CompletableFuture<LogicMatch> = tryRead(LogicTemplate.of(template))

    @JvmDefault
    fun tryAbsent(template: Term): CompletableFuture<LogicMatch> = tryAbsent(LogicTemplate.of(template))

    @JvmDefault
    fun tryTake(template: Term): CompletableFuture<LogicMatch> = tryTake(LogicTemplate.of(template))

    @JvmDefault
    fun read(template: Term): CompletableFuture<LogicMatch> = read(LogicTemplate.of(template))

    @JvmDefault
    fun tryReadTuple(template: Term): CompletableFuture<Optional<LogicTuple>> = tryReadTuple(LogicTemplate.of(template))

    @JvmDefault
    fun takeTuple(template: Term): CompletableFuture<LogicTuple> = takeTuple(LogicTemplate.of(template))

    @JvmDefault
    fun take(template: Term): CompletableFuture<LogicMatch> = take(LogicTemplate.of(template))

    @JvmDefault
    fun tryTakeTuple(template: Term): CompletableFuture<Optional<LogicTuple>> = tryTakeTuple(LogicTemplate.of(template))

    @JvmDefault
    fun absent(template: Term): CompletableFuture<LogicMatch> = absent(LogicTemplate.of(template))

    @JvmDefault
    fun tryAbsentTuple(template: Term): CompletableFuture<Optional<LogicTuple>> =
            tryAbsentTuple(LogicTemplate.of(template))

    @JvmDefault
    override fun String.toTuple(): LogicTuple = LogicTuple.of(this)

    @JvmDefault
    override fun String.toTemplate(): LogicTemplate = LogicTemplate.of(this)

    companion object {

        @JvmStatic
        fun local(name: String?, executorService: ExecutorService): LogicSpace =
                when (name) {
                    null -> LogicSpaceImpl(executor = executorService)
                    else -> LogicSpaceImpl(name, executorService)
                }

        @JvmStatic
        fun local(name: String?): LogicSpace = local(name, Engines.defaultEngine)

        @JvmStatic
        fun local(executorService: ExecutorService): LogicSpace = local(null, executorService)

        @JvmStatic
        fun local(): LogicSpace = local(null, Engines.defaultEngine)
    }
}
